src/os/bluestore/BlueStore.h

```diff
@@ -1084,38 +1086,137 @@ class BlueStore : public ObjectStore,
-    std::atomic<uint64_t> num_blobs = {0};
+    std::atomic<uint64_t> max = {0};
 
-    std::array<std::pair<ghobject_t, mono_clock::time_point>, 64> dumped_onodes;
+    CacheShard(CephContext* cct) : cct(cct), logger(nullptr) {}
+    virtual ~CacheShard() {}
+
+    void set_max(uint64_t max_) {
+      max = max_;
+    }
+
+    virtual void _trim_to(uint64_t max) = 0;
+    void _trim() {
+      _trim_to(max);
+    }
+    void trim() {
+      std::lock_guard l(lock);
+      _trim();    
+    }
+    void flush() {
+      std::lock_guard l(lock);
+     _trim_to(0);
+    }
+
+#ifdef DEBUG_CACHE
+    virtual void _audit(const char *s) = 0;
+#else
+    void _audit(const char *s) { /* no-op */ }
+#endif
+  };
+
+  /// A Generic onode Cache Shard 
+  struct OnodeCacheShard : public CacheShard {
+
+  public:
+    OnodeCacheShard(CephContext* cct) : CacheShard(cct) {}
+    static OnodeCacheShard *create(CephContext* cct, string type,
+                                   PerfCounters *logger) {
+      OnodeCacheShard *c = nullptr;
+      // Currently we only implement an LRU cache for onodes
+      c = new LruOnodeCacheShard(cct);
+      c->logger = logger;
+      return c;
+    }
+
+    virtual void _add(OnodeRef& o, int level) = 0;
+    virtual void _rm(OnodeRef& o) = 0;
+    virtual void _touch(OnodeRef& o) = 0;
+    virtual uint64_t _get_num() = 0;
+    virtual void add_stats(uint64_t *onodes) = 0;
+
+    bool empty() {
+      std::lock_guard l(lock);
+      return _get_num() == 0;
+    }
+  };
 
-    static Cache *create(CephContext* cct, string type, PerfCounters *logger);
+  /// An LRU onode Cache Shard
+  struct LruOnodeCacheShard : public OnodeCacheShard {
+    typedef boost::intrusive::list<
+      Onode,
+      boost::intrusive::member_hook<
+        Onode,
+        boost::intrusive::list_member_hook<>,
+        &Onode::lru_item> > list_t;
+    list_t lru;
+
+    LruOnodeCacheShard(CephContext *cct) : OnodeCacheShard(cct) {}
 
-    Cache(CephContext* cct) : cct(cct), logger(nullptr) {}
-    virtual ~Cache() {}
+    uint64_t _get_num() override {
+      return lru.size();
+    }
+    void _add(OnodeRef& o, int level) override {
+      (level > 0) ? lru.push_front(*o) : lru.push_back(*o);
+    }
+    void _rm(OnodeRef& o) override {
+      lru.erase(lru.iterator_to(*o));
+    }
+    void _touch(OnodeRef& o) override {
+      lru.erase(lru.iterator_to(*o));
+      lru.push_front(*o);
+    }
+    void _trim_to(uint64_t max) override;
 
-    virtual void _add_onode(OnodeRef& o, int level) = 0;
-    virtual void _rm_onode(OnodeRef& o) = 0;
-    virtual void _touch_onode(OnodeRef& o) = 0;
+    void add_stats(uint64_t *onodes) override {
+      std::lock_guard l(lock);
+      *onodes += lru.size();
+    }
+#ifdef DEBUG_CACHE
+    void _audit(const char *s) override;
+#endif
+  };
+
+  /// A Generic buffer Cache Shard
+  struct BufferCacheShard : public CacheShard {
+    std::array<std::pair<ghobject_t, mono_clock::time_point>, 64> dumped_onodes;
```

Comment:
Seems a little shady to me to leave this in the BufferCacheShard.  Should go back and look at that again.

Reply:
Also, I wonder if "Buffer Cache" is still the right name for this.

Reply:
i think 'buffer' is still accurate

Reply:
Moved!